Add RViz Marker DisplayTypes Tutorial (#4821)
ros2/ros2_documentation: source/Tutorials/Intermediate/RViz/Marker-Display-types/Marker-Display-types.rst

Marker: Display types
=====================

**Goal:** This tutorial explains the basic Marker types and how to use them.

**Tutorial level:** Intermediate

**Time:** 15 Minutes

.. contents:: Contents
   :depth: 2
   :local:

Background
----------
The Markers display allows programmatic addition of various primitive shapes to the 3D view by sending a
`visualization_msgs/msg/Marker <https://github.com/ros2/common_interfaces/blob/{DISTRO}/visualization_msgs/msg/Marker.msg>`_ or
`visualization_msgs/msg/MarkerArray <https://github.com/ros2/common_interfaces/blob/{DISTRO}/visualization_msgs/msg/MarkerArray.msg>`_ message.

.. image:: images/marker_overview.png

..
    This is a comment, the next line will be added to the file once the Markers-Sending-Basic-Shapes-CPP is merged to point to the start of the tutorial series.
    The :doc:`Marker: Sending Basic Shapes <../Markers-Sending-Basic-Shapes-CPP/Markers-Sending-Basic-Shapes-CPP>` that tutorial begins a series of tutorials on sending markers.

The Marker Message
------------------
1 Example Usage (C++)
^^^^^^^^^^^^^^^^^^^^^

> First we need to create a publisher and set a topic:

Reviewer: ```suggestion
First we will create a simple publisher node that publishes ``Marker`` messages from the ``visualization_messages`` package to the ``visualization_marker`` topic:
```

    marker.header.stamp = rclcpp::Clock().now();

    marker.ns = "basic_shapes";
    marker.id = 0;

    marker.type = visualization_msgs::msg::Marker::SPHERE;

    marker.action = visualization_msgs::msg::Marker::ADD;

    marker.pose.position.x = 0;
    marker.pose.position.y = 0;
    marker.pose.position.z = 0;
    marker.pose.orientation.x = 0.0;
    marker.pose.orientation.y = 0.0;
    marker.pose.orientation.z = 0.0;
    marker.pose.orientation.w = 1.0;

    marker.scale.x = 1.0;
    marker.scale.y = 1.0;
    marker.scale.z = 1.0;

    marker.color.r = 0.0f;
    marker.color.g = 1.0f;
    marker.color.b = 0.0f;
    marker.color.a = 1.0;   // Don't forget to set the alpha!

    // only if using a MESH_RESOURCE marker type:
    marker.mesh_resource = "package://pr2_description/meshes/base_v0/base.dae";

    marker.lifetime = rclcpp::Duration::from_nanoseconds(0);

    marker_pub->publish(marker);

There is also a `visualization_msgs/msg/MarkerArray <https://github.com/ros2/common_interfaces/blob/{DISTRO}/visualization_msgs/msg/MarkerArray.msg>`_ message, which lets you publish many markers at once.

2 Message Parameters
^^^^^^^^^^^^^^^^^^^^

> * ``ns``:

Reviewer: ```suggestion

The Marker message type is defined in `ROS 2 Common Interfaces <https://github.com/ros2/common_interfaces/tree/{DISTRO}/visualization_msgs/msg>_` package.
The messages in this package include comments that are helpful in understanding each of the fields in the message.

* ``ns``:
```
Where possible we should point users to message definitions, especially if they are reasonably documented.
Unfortunately the docs.ros.org doesn't include the actual message definitions. Github is the next best thing. We may want to use a macro for the ROS version in the linl.

    Namespace for these markers. This plus the id form a unique identifier.

* ``id``:

    Unique id assigned to this marker. It is your responsibility to keep these unique within your namespace.

* ``type``:

    Type of marker (Arrow, Sphere, ...). The available types are specified in the message definition.

* ``action``:

    0 = add/modify, 1 = (deprecated), 2 = delete, 3 = deleteall

* ``pose``:

    Pose marker, specified as x/y/z position and x/y/z/w quaternion orientation.

* ``scale``:

>     Scale of the marker. Applied before the position/orientation. A scale of [1,1,1] means the object will be 1m by 1m by 1m.

Reviewer 2: ```suggestion
    Scale of the marker. Applied before the position/orientation. A scale of [1, 1, 1] means the object will be 1m by 1m by 1m.
```

* ``color``:

>     Color of the object, specified as r/g/b/a, with values in the range of [0, 1]. Don't forget to set a or it will default to 0 and be invisible.

Reviewer: ```suggestion
    Color of the object, specified as r/g/b/a, with values in the range of [0, 1]. The, `a` or alpha value, denotes the opacity of the marker with 1 indicating opaque and 0 indicating completely transparent. The default value is 0, or completely transparent. **You must set the a value of your marker to a non-zero value or it will be transparent by default!**
```

* ``points``:

    Only used for markers of type ``Points``, ``Line strips``, and ``Line`` / ``Cube`` / ``Sphere`` -lists.

>     It's also used for the Arrow type, if you want to specify the arrow start and end points.

Reviewer: ```suggestion
    It's also used for the Arrow type, if you want to specify the arrow start and end points.
    This entry represents a list of `geometry_msgs/Point` types for the center or each marker object you would like rendered.  
```

* ``colors``:

>     Only used for markers that use the points member, specifies per-vertex color (no alpha yet).

Reviewer: ```suggestion
    This field is only used for markers that use the points member. This field specifies per-vertex color r/g/b/ color  (no alpha yet) for each entry in `points`.
```

* ``lifetime``:

>     A duration value used to automatically delete the marker after this period of time.

Reviewer: ```suggestion
    A `duration message value <https://docs.ros.org/en/ros2_packages/{DISTRO}/api/builtin_interfaces/interfaces/msg/Duration.html>`_ used to automatically delete the marker after this period of time.
```
Make the user's life easier. =)

    The countdown resets if another marker of the same ``namespace`` / ``id`` is received.

* ``frame_locked``:

>     Tells RViz to retransform the marker into the current location of the specified frame every update cycle.

Reviewer: Not your fault, but I wish this was a bit clearer.  The docs are pretty sparse.

Author: I changed it a bit. Mainly added a sentence, I hope this makes it clearer.

* ``text``:

    The text string used for the ``TEXT_VIEW_FACING`` marker type

* ``mesh_resource``:

>     The resource location for the ``MESH_RESOURCE`` marker type. Can be any mesh type supported by RViz (.stl or Ogre .mesh in 1.0, with the addition of COLLADA in 1.1).

Reviewer 2: ```suggestion
    The resource location for the ``MESH_RESOURCE`` marker type. Can be any mesh type supported by RViz (``.stl`` or Ogre ``.mesh`` in 1.0, with the addition of COLLADA in 1.1).
```

>     The format is the URI-form used by resource_retriever, including the package:// syntax.

Reviewer 2: link to `resource_retriever` ?

Line lists also have some special handling for scale: only ``scale.x`` is used and it controls the width of the line segments.

Note that ``pose`` is still used (the points in the line will be transformed by them), and the lines will be correct relative to the ``frame id`` specified in the header.

3.7 Cube List (CUBE_LIST=6)
~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. image:: images/CubeListMarker.png

A cube list is a list of cubes with all the same properties except their positions.
Using this object type instead of a `visualization_msgs/msg/MarkerArray <https://github.com/ros2/common_interfaces/blob/{DISTRO}/visualization_msgs/msg/MarkerArray.msg>`_ allows RViz to batch-up rendering,
which causes them to render much faster.
The caveat is that they all must have the same scale.

The ``points`` member of the `visualization_msgs/msg/Marker <https://github.com/ros2/common_interfaces/blob/{DISTRO}/visualization_msgs/msg/Marker.msg>`_ message is used for the position of each cube.

3.8 Sphere List (SPHERE_LIST=7)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. image:: images/SphereListMarker.png

A sphere list is a list of spheres with all the same properties except their positions.

> Using this object type instead of a `visualization_msgs/msg/MarkerArray <https://github.com/ros2/common_interfaces/blob/{DISTRO}/visualization_msgs/msg/MarkerArray.msg>`_ allows rviz to batch-up rendering,

Reviewer 2: ```suggestion
Using this object type instead of a `visualization_msgs/msg/MarkerArray <https://github.com/ros2/common_interfaces/blob/{DISTRO}/visualization_msgs/msg/MarkerArray.msg>`_ allows RViz to batch-up rendering,
```

which causes them to render much faster.
The caveat is that they all must have the same scale.

The ``points`` member of the `visualization_msgs/msg/Marker <https://github.com/ros2/common_interfaces/blob/{DISTRO}/visualization_msgs/msg/Marker.msg>`_ message is used for the position of each sphere.

Note that ``pose`` is still used (the ``points`` in the line will be transformed by them), and the lines will be correct relative to the ``frame id`` specified in the header.

3.9 Points (POINTS=8)
~~~~~~~~~~~~~~~~~~~~~

.. image:: images/PointsMarker.png

Uses the ``points`` member of the `visualization_msgs/msg/Marker <https://github.com/ros2/common_interfaces/blob/{DISTRO}/visualization_msgs/msg/Marker.msg>`_ message.

``Points`` have some special handling for scale: ``scale.x`` is point width, ``scale.y`` is point height

Note that ``pose`` is still used (the ``points`` in the line will be transformed by them), and the lines will be correct relative to the ``frame id`` specified in the header.

3.10 View-Oriented Text (TEXT_VIEW_FACING=9)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. image:: images/text_view_facing_marker.png

This marker displays text in a 3D spot in the world.

> The text always appears oriented correctly to the view. Uses the ``text`` field in the marker.

Reviewer: ```suggestion
The text always appears oriented correctly for the RViZ user to see the included text. Uses the ``text`` field in the marker.
```

Only ``scale.z`` is used. ``scale.z`` specifies the height of an uppercase "A".

3.11 Mesh Resource (MESH_RESOURCE=10)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. image:: images/mesh_resource_marker.png

Uses the ``mesh_resource`` field in the marker.

> Can be any mesh type supported by RViz (binary .stl or Ogre .mesh in 1.0, with the addition of COLLADA (.dae) in 1.1).

Reviewer 2: ```suggestion
Can be any mesh type supported by RViz (binary ``.stl`` or Ogre ``.mesh`` in 1.0, with the addition of COLLADA (.dae) in 1.1).
```

> The format is the URI-form used by resource_retriever, including the ``package://`` syntax.

Reviewer 2: link to `resource_retriever`?
